File: jbake-core/src/main/java/org/jbake/app/DocumentList.java
package org.jbake.app;

import com.arcadedb.query.sql.executor.Result;
import com.arcadedb.query.sql.executor.ResultSet;
import org.jbake.model.DocumentModel;

import java.util.*;

/**
 * Wraps an OrientDB document iterator into a model usable by
 * template engines.
 *
 * @author Cédric Champeau
 */
public class DocumentList<T> extends LinkedList<T> {

    public static DocumentList<DocumentModel> wrap(ResultSet docs) {
        DocumentList<DocumentModel> list = new DocumentList<>();
        while (docs.hasNext()) {
            Result next = docs.next();
            list.add(DBUtil.documentToModel(next));
        }
        docs.close();
        return list;
    }

}

File: jbake-core/src/main/java/org/jbake/app/DocumentList.kt
package org.jbake.app

import com.orientechnologies.orient.core.sql.executor.OResultSet
import org.jbake.model.DocumentModel
import java.util.*

/**
 * Wraps an OrientDB document iterator into a model usable by template engines.
 * Uses delegation instead of inheritance for better design.
 */
class DocumentList<T>(
    private val delegate: MutableList<T> = LinkedList()
) : MutableList<T> by delegate {

    fun push(element: T) = add(0, element)

    companion object {
        fun wrap(docs: OResultSet): DocumentList<DocumentModel> {
            val list = DocumentList<DocumentModel>()
            while (docs.hasNext()) {
                val next = docs.next()
                list.add(DbUtils.documentToModel(next))
            }
            docs.close()
            return list
        }
    }
}
